code/forms/UserForm.php

```diff
@@ -14,13 +14,15 @@ public function __construct(Controller $controller, $name = 'Form') {
 		$this->controller = $controller;
 		$this->setRedirectToFormOnValidationError(true);
 
-		parent::__construct(
-			$controller,
-			$name,
-			$this->getFormFields(),
-			$this->getFormActions(),
-			$this->getRequiredFields()
-		);
+		// Since these methods rely on extensions, replace
+		// them after parent::__construct has bootstrapped
+		parent::__construct($controller, $name, FieldList::create(), FieldList::create());
+		$this->setFields($this->getFormFields());
+		$this->setActions($this->getFormActions());
+		$this->setValidator($this->getRequiredFields());
```

Comment:
Oh, this is an amazingly annoying thing...

For extensions to work you MUST call parent::__construct before invoking methods which might touch extensions. This is because of bootstrapping in Object::__construct.

The problem is that $this->getFormFields(), and so on, each call $this->extend. These extension points, of course, were not working, so this is the workaround.